### PR TITLE
[Bots][Codeupdate][1471520]

### PR DESCRIPTION
**Pull request by Mahesh2516**
Updated the C# codes under [C#/.NET]. Please review.

### CODE REVIEW

**Review by KirtiPereira — CHANGES_REQUESTED** [review]
@Mahesh2516 - please do a thorough self-review. 
Example, continued usage of the word below.
"The code snippets in the [examples](#examples) section below are for..."

File: msteams-platform/bots/how-to/conversations/send-proactive-messages.md [hunk 1]
```diff
@@ -36,109 +36,101 @@ If you need to create a new conversation or conversation thread in a channel you
...
 
-Occasionally it may be necessary to proactively message users that have not installed or interacted with your app previously. For example, you want to use the [company communicator](~/samples/app-templates.md#company-communicator) to send messages to your entire organization. For this scenario you can use the Graph API to proactively install your app for your users, then cache the necessary values from the `conversationUpdate` event your app will receive upon install.
+Occasionally it may be necessary to proactively message users that have not installed or interacted with your app previously. For example, you want to use the [company communicator](~/samples/app-templates.md#company-communicator) to send messages to your entire organization. For this scenario you can use the Graph API to proactively install your app for your users, then cache the necessary values from the `conversationUpdate` event your app receives upon installation.
 
-You can only install apps that are in your organizational app catalogue, or the Teams app store.
+You can only install apps that are in your organizational app catalog, or the Teams app store.
```
Comment (KirtiPereira) [inline msteams-platform/bots/how-to/conversations/send-proactive-messages.md:88]:
```Suggestion
You can only install apps that are in your organizational app catalog or the Teams app store.
```

File: msteams-platform/bots/how-to/conversations/send-proactive-messages.md [hunk 2]
```diff
@@ -36,109 +36,101 @@ If you need to create a new conversation or conversation thread in a channel you
...
+When using proactive messaging to send notifications you need to make sure your users have a clear path to take common actions based on your notification and a clear understanding of why the notification occurred. Good notification messages generally include:
 
 * **What happened.** A clear indication of what happened to cause the notification.
-* **What was the result.** It should be clear what item/thing was updated to cause the notification.
+* **What was the result.** It must be clear what item or thing was updated to cause the notification.
 * **Who/what triggered it.** Who or what took action that caused the notification to be sent.
 * **What can users do in response.** Make it easy for your users to take actions based on your notifications.
-* **How can users opt out.** You need to provide a path for users to opt out of additional notifications.
+* **How can users opt-out.** You need to provide a path for users to opt-out of additional notifications.
```
Comment (KirtiPereira) [inline msteams-platform/bots/how-to/conversations/send-proactive-messages.md:79]:
```Suggestion
* **How can users opt out.** You must provide a path for users to opt out of additional notifications.
```

File: msteams-platform/bots/how-to/conversations/send-proactive-messages.md [hunk 3]
```diff
@@ -36,109 +36,101 @@ If you need to create a new conversation or conversation thread in a channel you
...
 * **What do you offer.** What can they do with your app? What value can you bring to them?
 * **What should they do next.** Invite them to try out a command, or interact with your app in some way.
 
-Remember, poor welcome messages can lead to users blocking your bot. You should spend plenty of time crafting your welcome messages, and iterate on them if they are not having the desired effect.
+Remember, poor welcome messages can lead to users blocking your bot. Spend plenty of time crafting your welcome messages, and iterate on them if they are not having the desired effect.
 
 ### Notification messages
 
-When using proactive messaging to send notifications you need to make sure your users have a clear path to take common actions based on your notification and a clear understanding of why the notification occurred. Good notification messages will generally include:
+When using proactive messaging to send notifications you need to make sure your users have a clear path to take common actions based on your notification and a clear understanding of why the notification occurred. Good notification messages generally include:
```
Comment (KirtiPereira) [inline msteams-platform/bots/how-to/conversations/send-proactive-messages.md:73]:
```Suggestion
When using proactive messaging to send notifications you must ensure your users have a clear path to take common actions based on your notification and a clear understanding of why the notification occurred. Good notification messages generally include:
```

File: msteams-platform/bots/how-to/conversations/send-proactive-messages.md [hunk 4]
```diff
@@ -36,109 +36,101 @@ If you need to create a new conversation or conversation thread in a channel you
...
 
 ## Best practices for proactive messaging
 
-Sending proactive messages to users can be a very effective way to communicate with your users. However, from their perspective this message can appear completely unprompted and, in the case of welcome messages, it will be the first time they've interacted with your app. As such, it is very important to use this functionality sparingly (don't spam your users) and to provide enough information to let users understand why they are being messaged.
+Sending proactive messages to users can be a very effective way to communicate with your users. However, from their perspective, this message can appear completely unprompted, and in the case of welcome messages, it is the first time they have interacted with your app. As such, it is very important to use this functionality sparingly (don't spam your users) and to provide enough information to let users understand why they are being messaged.
```
Comment (KirtiPereira) [inline msteams-platform/bots/how-to/conversations/send-proactive-messages.md:59]:
```Suggestion
Sending proactive messages to users is a very effective way to communicate with your users. However, from their perspective, this message can appear completely unprompted, and in the case of welcome messages, it is the first time they have interacted with your app. Therefore, it is very important to use this functionality sparingly, don't spam your users, and to provide enough information to let users understand why they are being messaged.
```

File: msteams-platform/bots/how-to/conversations/send-proactive-messages.md [hunk 5]
```diff
@@ -36,109 +36,101 @@ If you need to create a new conversation or conversation thread in a channel you
...
+Once you have the user/channel information, you'll need to create the conversation if it doesn't already exist (or you don't know the `conversationId`). You must only create the conversation once; make sure you store the `conversationId` value or `conversationReference` object to use in the future.
 
 ## Get the conversation ID
 
-Once the conversation has been created, you will use either the `conversationReference` object or the `conversationId` and the `tenantId` to send the message. You can get this Id by either creating the conversation, or storing it from any Activity sent to you from that context. Make certain that you store this Id.
+Once the conversation has been created, you must use either the `conversationReference` object or the `conversationId` and the `tenantId` to send the message. You can get this ID by either creating the conversation, or storing it from any Activity sent to you from that context. Make certain that you store this Id.
```
Comment (KirtiPereira) [inline msteams-platform/bots/how-to/conversations/send-proactive-messages.md:51]:
```Suggestion
After the conversation is created, use either the `conversationReference` object or `conversationId` and `tenantId` to send the message. You can get this ID by either creating the conversation or storing it from any Activity sent to you from that context. Make certain that you store this ID.
```

File: msteams-platform/bots/how-to/conversations/send-proactive-messages.md [hunk 6]
```diff
@@ -36,109 +36,101 @@ If you need to create a new conversation or conversation thread in a channel you
...
 
-The `userId` is unique to your bot Id and a particular user, you cannot re-use them between bots. The `channelId` is global, however your bot _must_ be installed in the team before you can send a proactive message to a channel.
+The `userId` is unique to your bot Id and a particular user, you cannot re-use them between bots. The `channelId` is global, however, your bot must be installed in the team before you can send a proactive message to a channel.
 
 ## Create the conversation
 
-Once you have the user/channel information, you'll need to create the conversation if it doesn't already exist (or you don't know the `conversationId`). You should only create the conversation once; make sure you store the `conversationId` value or `conversationReference` object to use in the future.
+Once you have the user/channel information, you'll need to create the conversation if it doesn't already exist (or you don't know the `conversationId`). You must only create the conversation once; make sure you store the `conversationId` value or `conversationReference` object to use in the future.
```
Comment (KirtiPereira) [inline msteams-platform/bots/how-to/conversations/send-proactive-messages.md:47]:
```Suggestion
After you have the user or channel information, you'll need to create the conversation if it doesn't already exist or you don't know the `conversationId`. You must only create the conversation once and make sure you store the `conversationId` value or `conversationReference` object to use in the future.
```

File: msteams-platform/bots/how-to/conversations/send-proactive-messages.md [hunk 7]
```diff
@@ -36,109 +36,101 @@ If you need to create a new conversation or conversation thread in a channel you
...
-1. Every Activity your bot receives will contain the necessary information.
+1. Every Activity your bot receives must contain the necessary information.
 
-Regardless of how you gain the information, you'll need to store the `tenantId` and either the `userId` or `channelId` in order to create a new conversation. You can also use the `teamId` to create a new conversation thread in the general/default channel of a team.
+Regardless of how you gain the information, you'll need to store the `tenantId` and either the `userId` or `channelId` to create a new conversation. You can also use the `teamId` to create a new conversation thread in the general or default channel of a team.
 
-The `userId` is unique to your bot Id and a particular user, you cannot re-use them between bots. The `channelId` is global, however your bot _must_ be installed in the team before you can send a proactive message to a channel.
+The `userId` is unique to your bot Id and a particular user, you cannot re-use them between bots. The `channelId` is global, however, your bot must be installed in the team before you can send a proactive message to a channel.
```
Comment (KirtiPereira) [inline msteams-platform/bots/how-to/conversations/send-proactive-messages.md:43]: Id --> ID

File: msteams-platform/bots/how-to/conversations/send-proactive-messages.md [hunk 8]
```diff
@@ -36,109 +36,101 @@ If you need to create a new conversation or conversation thread in a channel you
 1. When a new user is added to a context where your app is installed, you'll receive a [`onMembersAdded` Activity](~/bots/how-to/conversations/subscribe-to-conversation-events.md).
 1. You can retrieve the [list of channels](~/bots/how-to/get-teams-context.md) in a team your app is installed.
 1. You can retrieve the [list of members](~/bots/how-to/get-teams-context.md) of a team your app is installed.
-1. Every Activity your bot receives will contain the necessary information.
+1. Every Activity your bot receives must contain the necessary information.
```
Comment (KirtiPereira) [inline msteams-platform/bots/how-to/conversations/send-proactive-messages.md:39]:
````Suggestion
Every Activity your bot receives must contain the required information.
```

File: msteams-platform/bots/how-to/conversations/send-proactive-messages.md [hunk 9]
```diff
@@ -15,7 +15,7 @@ A proactive message is any message sent by a bot that is not in direct response
...
+For your bot to send a proactive message, it must have access to the user, group chat, or team that you wish to send the message to. For a group chat or team, this means the app that contains your bot must be installed to that location first. You can [proactively install your app using Graph](#proactively-install-your-app-using-graph) in a team if necessary or use an [app policy](/microsoftteams/teams-custom-app-policies-and-settings) to push apps out to teams and users in your tenant. For users, your app either needs to be installed for that user or your user needs to be part of a team where your app is installed.
```
Comment (KirtiPereira) [inline msteams-platform/bots/how-to/conversations/send-proactive-messages.md:18]: The following sentence is not clear. 
For users, your app either needs to be installed for that user or your user needs to be part of a team where your app is installed.
Should it be, For the users to use the app, it must either be installed for the user or the user must be part of a team where your app is installed.

```Suggestion
For your bot to send a proactive message, it must have access to the user, group chat, or team that you want to send the message to. For a group chat or team, this means the app that contains your bot must be installed to that location first. You can [proactively install your app using Graph](#proactively-install-your-app-using-graph) in a team if necessary or use an [app policy](/microsoftteams/teams-custom-app-policies-and-settings) to push apps out to teams and users in your tenant. For users, your app either must be installed for that user or your user must be part of a team where your app is installed.
```

**Review by KirtiPereira — CHANGES_REQUESTED** [review]
@Mahesh2516 - please verify and commit the suggestions.

File: msteams-platform/bots/how-to/conversations/send-proactive-messages.md [hunk 10]
```diff
@@ -15,9 +15,9 @@ A proactive message is any message sent by a bot that is not in direct response
...
+For your bot to send a proactive message, it must have access to the user, group chat, or team that you want to send the message to. For a group chat or team, this means the app that contains your bot must be installed to that location first. You can [proactively install your app using Graph](#proactively-install-your-app-using-graph) in a team if necessary or use an [app policy](/microsoftteams/teams-custom-app-policies-and-settings) to push apps out to teams and users in your tenant. For users, your app either must be installed for that user or your user must be part of a team where your app is installed.
```
Comment (KirtiPereira) [inline msteams-platform/bots/how-to/conversations/send-proactive-messages.md:18]: What is "that location"?
```Suggestion
For your bot to send a proactive message, it must have access to the user, group chat, or team that you want to send the message to. For a group chat or team, this means the app that contains your bot must be installed to that location first. You can [proactively install your app using Graph](#proactively-install-your-app-using-graph) in a team, if required or use an [app policy](/microsoftteams/teams-custom-app-policies-and-settings) to push apps out to teams and users in your tenant. For users, your app either must be installed for the user or your user must be part of a team where your app is installed.
```

File: msteams-platform/bots/how-to/conversations/send-proactive-messages.md [hunk 11]
```diff
@@ -15,9 +15,9 @@ A proactive message is any message sent by a bot that is not in direct response
...
 
-Sending a proactive message is different than sending a regular message in that you won't have an active `turnContext` to use for a reply. You may also need to create the conversation (for example a new one-to-one chat, or a new conversation thread in a channel) before sending the message. You cannot create a new group chat or a new channel in a team with proactive messaging.
+Sending a proactive message is different than sending a regular message. In that, there is no active `turnContext` to use for a reply. You may also need to create the conversation (for example a new one-to-one chat, or a new conversation thread in a channel) before sending the message. You cannot create a new group chat or a new channel in a team with proactive messaging.
```
Comment (KirtiPereira) [inline msteams-platform/bots/how-to/conversations/send-proactive-messages.md:20]: What is "In that,..."?
```Suggestion
Sending a proactive message is different than sending a regular message. In that, there is no active `turnContext` to use for a reply. You may also need to create the conversation before sending the message. For example, a new one-to-one chat or a new conversation thread in a channel. You cannot create a new group chat or a new channel in a team with proactive messaging.
```

File: msteams-platform/bots/how-to/conversations/send-proactive-messages.md [hunk 12]
```diff
@@ -26,119 +26,111 @@ At a high level the steps you'll need to complete to send a proactive message ar
 1. [Get the conversation ID](#get-the-conversation-id).
 1. [Send the message](#send-the-message).
 
-The code snippets in the [examples](#examples) section below are for creating a one-to-one conversation, see the [references](#references) section for links to complete working samples for both one-to-once conversations and group/channels.
+The code snippets in the [examples](#examples) section are for creating a one-to-one conversation, see the [references](#references) section for links to complete working samples for both one-to-one conversations and group or channels.
```
Comment (KirtiPereira) [inline msteams-platform/bots/how-to/conversations/send-proactive-messages.md:29]:
```Suggestion
The code snippets in the [examples](#examples) section are for creating a one-to-one conversation. For links to complete working samples for both one-to-one conversations and group or channels , see [references](#references).
```

File: msteams-platform/bots/how-to/conversations/send-proactive-messages.md [hunk 13]
```diff
@@ -26,119 +26,111 @@ At a high level the steps you'll need to complete to send a proactive message ar
...
 
-If you need to create a new conversation or conversation thread in a channel you'll first need the right ID to create the conversation. You can receive/retrieve this ID in multiple ways:
+To create a new conversation or conversation thread in a channel, you need the correct ID. You can receive or retrieve this ID in multiple ways:
 
 1. When your app is installed in any particular context, you'll receive a [`onMembersAdded` Activity](~/bots/how-to/conversations/subscribe-to-conversation-events.md).
 1. When a new user is added to a context where your app is installed, you'll receive a [`onMembersAdded` Activity](~/bots/how-to/conversations/subscribe-to-conversation-events.md).
 1. You can retrieve the [list of channels](~/bots/how-to/get-teams-context.md) in a team your app is installed.
 1. You can retrieve the [list of members](~/bots/how-to/get-teams-context.md) of a team your app is installed.
-1. Every Activity your bot receives will contain the necessary information.
+Every Activity your bot receives must contain the required information.
```
Comment (KirtiPereira) [inline msteams-platform/bots/how-to/conversations/send-proactive-messages.md:39]: Where is the numbering?
```Suggestion
1. Every Activity your bot receives must contain the required information.
```

File: msteams-platform/bots/how-to/conversations/send-proactive-messages.md [hunk 14]
```diff
@@ -26,119 +26,111 @@ At a high level the steps you'll need to complete to send a proactive message ar
...
 
-The `userId` is unique to your bot Id and a particular user, you cannot re-use them between bots. The `channelId` is global, however your bot _must_ be installed in the team before you can send a proactive message to a channel.
+The `userId` is unique to your bot Id and a particular user, you cannot re-use them between bots. The `channelId` is global, however, your bot must be installed in the team before you can send a proactive message to a channel.
 
 ## Create the conversation
 
-Once you have the user/channel information, you'll need to create the conversation if it doesn't already exist (or you don't know the `conversationId`). You should only create the conversation once; make sure you store the `conversationId` value or `conversationReference` object to use in the future.
+After you have the user or channel information, you'll need to create the conversation if it doesn't already exist or you don't know the `conversationId`. You must only create the conversation once and make sure you store the `conversationId` value or `conversationReference` object to use in the future.
```
Comment (KirtiPereira) [inline msteams-platform/bots/how-to/conversations/send-proactive-messages.md:47]:
```Suggestion
After you have the user or channel information, you need to create the conversation if it doesn't already exist or you don't know the `conversationId`. You must only create the conversation once and make sure you store the `conversationId` value or `conversationReference` object to use in the future.
```

**Review by KirtiPereira — CHANGES_REQUESTED** [review]
Verify and update the document.

File: msteams-platform/bots/how-to/conversations/send-proactive-messages.md [hunk 15]
```diff
@@ -26,119 +26,111 @@ At a high level the steps you'll need to complete to send a proactive message ar
...
 
-You can only install apps that are in your organizational app catalogue, or the Teams app store.
+You can only install apps that are in your organizational app catalog or the Teams app store.
 
-See [Install apps for users](/graph/teams-proactive-messaging) in the Graph documentation and [Proactive bot installation and messaging in Teams with Microsoft Graph](../../../graph-api/proactive-bots-and-messages/graph-proactive-bots-and-messages.md). There is also a [Microsoft .NET framework sample](https://github.com/microsoftgraph/contoso-airlines-teams-sample/blob/283523d45f5ce416111dfc34b8e49728b5012739/project/Models/GraphService.cs#L176)  on the GitHub platform.
+See [Install apps for users](/graph/teams-proactive-messaging) in the Graph documentation and [Proactive bot installation and messaging in Teams with Microsoft Graph](../../../graph-api/proactive-bots-and-messages/graph-proactive-bots-and-messages.md). There is also a [Microsoft .NET framework sample](https://github.com/microsoftgraph/contoso-airlines-teams-sample/blob/283523d45f5ce416111dfc34b8e49728b5012739/project/Models/GraphService.cs#L176) on the GitHub platform.
```
Comment (Trinetra-MSFT) [inline msteams-platform/bots/how-to/conversations/send-proactive-messages.md:90]: [Install apps for users](https://docs.microsoft.com/en-us/graph/api/userteamwork-post-installedapps?view=graph-rest-1.0&tabs=http)

Comment (Trinetra-MSFT) [reply msteams-platform/bots/how-to/conversations/send-proactive-messages.md:90]: Broken link

**Review by Trinetra-MSFT — COMMENTED** [review]
Please test the links in doc

File: msteams-platform/bots/how-to/conversations/send-proactive-messages.md (hunk 15 above)
Comment (Mahesh2516) [reply msteams-platform/bots/how-to/conversations/send-proactive-messages.md]: Added the absolute path for **Install apps for users**.

**Review by Trinetra-MSFT — DISMISSED** [review]

**Review by KirtiPereira — APPROVED** [review]
Approve

File: msteams-platform/bots/how-to/conversations/send-proactive-messages.md [hunk 16]
```diff
@@ -243,19 +235,19 @@ You must supply the user ID and the tenant ID. If the call succeeds, the API ret
 
 ## References
 
-The official proactive messaging samples are listed below.
+The official proactive messaging samples are as follows:
```
Comment (surbhigupta12) [inline msteams-platform/bots/how-to/conversations/send-proactive-messages.md:238]: @KirtiPereira and @Mahesh2516,

Please update the sample code part. Change the heading "References"

**Review by surbhigupta12 — APPROVED** [review]
Approving the PR.

**Review by KirtiPereira — APPROVED** [review]
approved

**Pull request merged**